[24.1] Allow drag and drop for collection elements

ahmedhamidawan opened the pull request:
We extract the `HDAObject` from the `DCESummary` to be able to drag and drop it. For now, this works on dragging and dropping items to histories, since that process only needs the dataset's id. However, this doesn't work for `DCECollections` because they are `dataset_collection`s and not `HDCA`s.

## Drag and drop DCEs into Tool forms:
Fixes https://github.com/galaxyproject/galaxy/issues/18692

https://github.com/user-attachments/assets/993b6ecc-0e13-422e-a21d-bd0dfecbf30a

## Drag and drop DCEs to other histories:
Note that this won't work for `DatasetCollectionElement`s that are of type `dataset_collection`. `DCEDataset`s can be moved:

https://github.com/user-attachments/assets/4e34f5a2-a0fa-4b81-8fad-24a1c469f8ca



## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

ahmedhamidawan committed:
[24.1] Allow drag and drop for collection elements

We extract the `HDAObject` from the `DCESummary` to be able to drag and drop it.
For now, this works on dragging and dropping items to histories, since that process only needs the dataset's id. However, this doesn't work on tool forms; as we would need to get the full `HDASummary` to do that.

ahmedhamidawan committed:
show name in draggable for `DCE`

ahmedhamidawan committed:
store DCEs as dragData, and check for `element_type` within `onDrop()`

for now, only `getDragData` for `DCEDataset`s

davelopez commented on client/src/api/datasets.ts:
```diff
@@ -67,7 +67,7 @@ export async function purgeDataset(datasetId: string) {
 
 const datasetCopy = fetcher.path("/api/histories/{history_id}/contents/{type}s").method("post").create();
-type HistoryContentsArgs = FetchArgType<typeof datasetCopy>;
+export type HistoryContentsArgs = FetchArgType<typeof datasetCopy>;
```
Just a heads up, this is going to create a conflict when merging forward since `FetchArgType` no longer exists in `dev`.
What I recommend is to leave this line as it was and then add at the bottom of this file the 2 types that you really need in  `HistoryPanel.vue`.

Something like this:
```ts
export type HistoryContentType = components["schemas"]["HistoryContentType"];
export type HistoryContentSource = components["schemas"]["HistoryContentSource"];
```

ahmedhamidawan replied:
Makes sense, thank you! I will try to keep in mind such cases where something could create conflicts merging forward 👍

ahmedhamidawan committed:
export explicit types for content source and type

Co-authored-by: David López <46503462+davelopez@users.noreply.github.com>

ahmedhamidawan committed:
create a reusable `historyDragDrop` composable

This unifies the drag and drop operations in `HistoryPanel` as well as the drop area in the `MultipleViewList`.

ahmedhamidawan committed:
use a computed to track changed history for drag drop

ahmedhamidawan committed:
allow DCEs to be dropped to tool form

mvdbeek commented on client/src/composables/historyDragDrop.ts:
```diff
@@ -0,0 +1,173 @@
+                    type = "dataset";
+                    id = item.object.id;
+                } else {
+                    throw new Error(`Invalid item type${item.element_type ? `: ${item.element_type}` : ""}`);
```
There was an error handler in https://github.com/galaxyproject/galaxy/pull/18699/files#diff-68e87a439364cbb0acfc54a106c0830012c8005d4288102ba12a57fc7ffcedd6L408, you might want to re-instate that. This btw can happen if the collection isn't populated yet, so we'll definitely run into this from time to time.

mvdbeek replied:
I see you added a handler in https://github.com/galaxyproject/galaxy/pull/18699/files#diff-ad4aa7272cdb314a83cba0349216030549831df46d0113ef0d5b60a67d72a26aR160 ... I think it might be better to do error handling like this in the caller, so context-specific actions can be taken (e.g. Toast vs inline errors in this example). Also the caller might want to know if the action succeeded, which it will always if you catch the error.

mvdbeek commented on client/src/composables/historyDragDrop.ts:
```diff
@@ -0,0 +1,173 @@
+                } else {
+                    throw new Error(`Invalid item type${item.element_type ? `: ${item.element_type}` : ""}`);
+                }
+                await copyDataset(id, historyId, type, dataSource);
```
Maybe you'd want to loop over the items to drag first and make sure none of them run into https://github.com/galaxyproject/galaxy/pull/18699/files#r1734725855 ? Otherwise it might be a little messy if a fraction of items is copied and another is not.

ahmedhamidawan replied:
I have moved the actual copy operation to another loop, so that if errors are caught for any item, we never copy anything from then on.

From what I understand, if there is any error with:
- `historyStore.createNewHistory();`
- `!historyId`
- Invalid item type
- `copyDataset`
- `loadHistoryById`

the error would be handled and Toast-ed in `useHistoryDragDrop.onDrop()`.

mvdbeek reviewed (commented):
Nice deduplication, thank you!

ahmedhamidawan committed:
move copy operation to another loop to only copy when no errors

ahmedhamidawan committed:
only load/pin history if items were copied
